frontend/server/src/main/java/org/pytorch/serve/wlm/ModelManager.java

```diff
@@ -152,6 +158,35 @@ private ModelArchive createModelArchive(
         return archive;
     }
 
+    private void setupModelDependencies(Model model) throws IOException, InterruptedException {
+        String requirementsFile =
+                model.getModelArchive().getManifest().getModel().getRequirementsFile();
+
+        if (configManager.getInstallPyDepPerModel() && requirementsFile != null) {
+            Path requirementsFilePath =
+                    Paths.get(model.getModelDir().getAbsolutePath(), requirementsFile);
+
+            String pythonRuntime = EnvironmentUtils.getPythonRunTime(model);
+
+            String packageInstallCommand =
+                    pythonRuntime
+                            + " -m pip install -U -t "
+                            + model.getModelDir().getAbsolutePath()
+                            + " -r "
+                            + requirementsFilePath; // NOPMD
+
+            String[] envp =
+                    EnvironmentUtils.getEnvString(configManager.getModelServerHome(), null, null);
+            Process process =
+                    Runtime.getRuntime()
```

Comment:
@harshbafna How are failure cases handled? eg wrong name of package in the requirements.txt or a package that takes very long to download from internet / fails to download. How are those errors handled? Further errors need to be handled gracefully and make sure that rest of TorchServe / workers do not crash.

And for models with multiple worker threads, the download should happen only once

Reply:
@chauhang 

> How are failure cases handled? eg wrong name of package in the requirements.txt 
Added error handling in case the `pip` install fails. Added [corresponding UT](https://github.com/pytorch/serve/pull/462/files#diff-fd2592adf5a38db5711d59c10dfe36bbR652) as well.

> a package that takes very long to download from internet / fails to download

The model registration request is synchronous and there is no defined timeout. In case the download fails, then the pip command will fail and TorchServe will return an error message to the client.

> And for models with multiple worker threads, the download should happen only once

The installation of the custom packages happens at the time of model registration in a temporary directory where the model archive is extracted. This path is then added to the PYTHONPATH and is passed on to the workers at the time of worker creation.